### DISCUSSION TITLE
Volume change

### DISCUSSION THREAD

**Discussion by ophirpeleg** [dtcooper/raspotify#555] [Ideas]
Hey guys,
My RASPI 3 is connected to speakers with the headphone jack, and I found a way to control the volume. It was delayed when I changed the volume via the Spotify app on my PC. 
Now, I connected an encoder with a button (KY-040) and controlled the volume physically.
short-press - mute / unmute
long-press - set to 0dB

Attaching here is the Python code: 
```

import board
import digitalio
import os
import os.path
import time


def millis():
    return time.monotonic() * 1000


def long_press():
    os.system(f"amixer -c 0 set Headphone 0")

os.system(f"amixer -c 0 set Headphone unmute")
mute = False

dirPin = digitalio.DigitalInOut(board.D2)
pushPin = digitalio.DigitalInOut(board.D17)
stepPin = digitalio.DigitalInOut(board.D3)

dirPin.direction = digitalio.Direction.INPUT
pushPin.direction = digitalio.Direction.INPUT
stepPin.direction = digitalio.Direction.INPUT
volumeInterval = 50

dirPin.pull = digitalio.Pull.UP
stepPin.pull = digitalio.Pull.UP
pushPin.pull = digitalio.Pull.UP
previousValue = True

print("Welcome")
print(f"{stepPin.value} step pin value, {dirPin.value} dir pin value")
print(f"push pin value: {pushPin.value}")
is_on = True

while is_on:
    if pushPin.value == 0:
        pressTime = millis()
        time.sleep(0.2)
        longPress = False

        while pushPin.value == 0:
            if millis() - pressTime > 1000 and not longPress:
                longPress = True
                long_press()

        if not longPress:  # short press define
            if mute:
                os.system(f"amixer -c 0 set Headphone unmute")
                mute = False
            else:
                os.system(f"amixer -c 0 set Headphone mute")
                mute = True

    if previousValue != stepPin.value:
        if stepPin.value == False:
            if dirPin.value == False:
                print("Right")
                os.system(f"amixer -c 0 set Headphone {volumeInterval}+")
            else:
                print("Left")
                os.system(f"amixer -c 0 set Headphone {volumeInterval}-")
        previousValue = stepPin.value


```

**Comment by JasonLG1979** [comment]
Very cool but I think that it must be pointed out as a disclaimer that the software volume control in librespot is superior in every way sound quality wise then any form of hardware volume control. If sound quality is your goal don't use hardware volume control.

**Reply by ophirpeleg** [reply]
Is there another way to accomplice volume change without sacrificing sound quality? 
and another idea I had is to add OLED - currently displaying volume and status (if muted or not), is there a way to get what songs are playing with raspotify terminal command?

**Reply by JasonLG1979** [reply]
> Is there another way to accomplice volume change without sacrificing sound quality?

No, not really. The best volume control for librespot is it's own software control. All internal audio processing done in librespot is done with 64 bit floating point precision with dithering when applicable (when using 16 or 24 bit output). DAC's are basically black boxes for the most part. It's impossible to know how they actually do volume control most of the time. It's also incorrect to conflate hardware volume control with analog volume control. 99.999% of the time hardware volume control just means that you're using the DAC's internal software volume control that may very well be inferior to librespot's software control.

For the headphone out on a Pi3 in particular you DO NOT want to use hardware volume control. The headphone out on Pi's is garbage. It's not even a proper DAC it's a crude PWM with a super basic analog filter. It's really not suitable for music at all. They themselves acknowledge as much and recommend leaving the hardware volume at 100% and using software volume control in whatever is making noise. A $10 USB audio dongle would be a HUGE upgrade in sound quality.

> and another idea I had is to add OLED - currently displaying volume and status (if muted or not), is there a way to get what songs are playing with raspotify terminal command?

No. librespot is strictly a player. You can get silly and enable verbose logging and parse some metadata info from the logs but it's much easier to just use one of the many other projects designed for interacting with the public Spotify web API.

**Reply by JasonLG1979** [reply]
Correction:

You can use an on-event script to get some info but it's very limited and for security reasons it's not very easy with the default Raspotify service. You would basically have to write your own systemd service and disable the Raspotify service.

An example on-event python script might look something like this:

```python3
import os

if os.environ['PLAYER_EVENT'] == 'changed':
    old_track_id = os.environ['OLD_TRACK_ID']
    new_track_id = os.environ['TRACK_ID']
    # do stuff

elif os.environ['PLAYER_EVENT'] == 'started':
    track_id = os.environ['TRACK_ID']
    # do suff

elif os.environ['PLAYER_EVENT'] == 'stopped':
    track_id = os.environ['TRACK_ID']
    # do stuff
   
elif os.environ['PLAYER_EVENT'] == 'playing':
    track_id = os.environ['TRACK_ID']
    track_duration_ms = os.environ['DURATION_MS']
    track_position_ms = os.environ['POSITION_MS']
    # do stuff

elif os.environ['PLAYER_EVENT'] == 'paused':
    track_id = os.environ['TRACK_ID']
    track_duration_ms = os.environ['DURATION_MS']
    track_position_ms = os.environ['POSITION_MS']
    # do stuff

elif os.environ['PLAYER_EVENT'] == 'preloading':
    track_id = os.environ['TRACK_ID']
    # do stuff

elif os.environ['PLAYER_EVENT'] == 'volume_set':
    volume = os.environ['VOLUME']
    # do stuff
```

You could then use that info to get metadata info from some other lib.

**Reply by JasonLG1979** [reply]
In the above example `os.environ['TRACK_ID']` is the Spotify track ID.

**Reply by JasonLG1979** [reply]
For reference:
https://github.com/librespot-org/librespot/blob/master/src/player_event_handler.rs

**Reply by giggywithit** [reply]
I'm looking for the best way to reset the volume after a period of idle, no playback. What happens is that I will listen at fairly high levels and then sometimes stop the playback and go do something. Upon returning, sometimes ill play another track and it'l be left at those high volume levels and blast when I am not prepared for it. For the moment, the only solution is to restart raspotify after a period of idle, in which I have set a start volume at a lower level. 

Thoughts?

**Reply by klockeph** [reply]
I'm on a similar quest right now (want to turn on/off the AV Receiver when librespot starts/ends playing, and relay volume control).

The systemd service override is not too bad (created by `sudo systemctl edit raspotify`):
```
[Service]
ProtectHome=false  #not required if you put your event handler somewhere else than your home directory
DeviceAllow=/dev/cec0 # for cec control

SupplementaryGroups=audio video # /dev/cec0 is owned by video group
ExecStart=
ExecStart=/usr/bin/librespot --onevent /home/philipp/librespotify/event_handler.sh
```

My event_handler script (WIP) looks somewhat like this:
```
#!/bin/bash
set -e

case $PLAYER_EVENT in
  "session_connected")
    echo 'on 5' | cec-client -s -d 1
    sleep 5 # wait for the AV receiver to turn on
    echo 'as' | cec-client -s -d 1
    ;;
  "session_disconnected")
    echo 'standby 5' | cec-client -s -d 1
    ;;
esac
```
However, I fear that `librespot` is quite unhappy when you use the HDMI port it uses for audio output for CEC... For starters this whole setup doesn't work. If I manually perform the `cec` commands before the session connects it works; but with the script I get no audio output.

Furthermore adding the volume control events results in errors from cec-client:
```
echo 'volup' | cec-client -s -d 1
opening a connection to the CEC adapter...
volume up: 7F
ERROR:   [            2003]     CLinuxCECAdapterCommunication::Write - ioctl CEC_TRANSMIT failed - tx_status=00 errno=1
ERROR:   [            2003]     CLinuxCECAdapterCommunication::Write - ioctl CEC_TRANSMIT failed - tx_status=00 errno=1
```
I couldn't find information on this error yet, but I assume it's because librespot is playing so cec-lib can't write to the hdmi port :/
